File: leetcode/021_Merge_Two_Sorted_Lists/main.c
#include <stdio.h>
#include <stdlib.h>
/*
 *Input: 1->2->4, 1->3->4
  Output: 1->1->2->3->4->4
 */

struct ListNode {
     int val;
     struct ListNode *next;
};
struct ListNode* mergeTwoLists(struct ListNode* l1, struct ListNode* l2);
void addNodes(int arr[],struct ListNode* node,int arrSize);
void freeNodes(struct ListNode* node);
void visitNodes(struct ListNode* node);

int main(void)
{
	int L1_Arr1[]={2};
	int L2_Arr2[]={1};
	//int L1_Arr1[]={};
	//int L2_Arr2[]={};


	struct ListNode L1;
	struct ListNode L2;
	struct ListNode *L3;

	int sizeA1=sizeof(L1_Arr1)/sizeof(int);
	int sizeA2=sizeof(L2_Arr2)/sizeof(int);
	addNodes(L1_Arr1,&L1,sizeA1);
	addNodes(L2_Arr2,&L2,sizeA2);
	
	visitNodes(&L1);
	visitNodes(&L2);

	L3=mergeTwoLists(&L1, &L2);

	visitNodes(L3);
	freeNodes(&L1);
	freeNodes(&L2);
	freeNodes(L3);

	return 0;
}
void freeNodes(struct ListNode* node){
	//free from second node
	struct ListNode* currNode=NULL;	
	struct ListNode* lastNode=NULL;	

	if(node->next==NULL)
		return;

	currNode=node->next;


	while(currNode->next!=NULL){
		lastNode=currNode;
		currNode=currNode->next;
		free(lastNode);
	}
		
	free(currNode);
	
}
void addNodes(int arr[],struct ListNode* node, int arrSize){

	struct ListNode* currNode=node;	
	struct ListNode* nextNode=NULL;	
	for (int i=0;i<arrSize;i++){
		currNode->val=arr[i];
		if (i==arrSize-1){
			currNode->next=NULL;	
		}else{
			nextNode=malloc(sizeof(struct ListNode));
			currNode->next=nextNode;
			currNode=currNode->next;		
		}
	}
}
void visitNodes(struct ListNode* node){
	struct ListNode* currNode=node;	
	printf("%d->", currNode->val);

	while(currNode->next!=NULL){
		currNode=currNode->next;
		printf("%d->", currNode->val);
	}

	printf("\n");
}
struct ListNode* mergeTwoLists(struct ListNode* l1, struct ListNode* l2) {
	struct ListNode* currL1=l1;	
	struct ListNode* currL2=l2;	
	struct ListNode* L3=NULL;
	struct ListNode* origin_L3=NULL;
	int value=0;

	if (l1==NULL && l2==NULL){
		    return NULL;
	}

	L3=(struct ListNode *)malloc(sizeof(struct ListNode));
	origin_L3=L3;
	L3->next=NULL;

	while(currL1!=NULL || currL2!=NULL){
		if(currL2==NULL){
			value=currL1->val;
			if(currL1->next!=NULL){
				currL1=currL1->next;
			}else{
				currL1=NULL;
			}		
		}
		else if(currL1==NULL)
		{
			value=currL2->val;
			if(currL2->next!=NULL){
				currL2=currL2->next;
			}else{
				currL2=NULL;
			}
	
		}
		else if(currL1->val<=currL2->val)
		{
			value=currL1->val;
			if(currL1->next!=NULL){
				currL1=currL1->next;
			}else{
				currL1=NULL;
			}	
		}
		else{
			value=currL2->val;
			if(currL2->next!=NULL){
				currL2=currL2->next;
			}else{
				currL2=NULL;
			}
		}
	
		L3->val=value;
		if(currL1!=NULL || currL2!=NULL){
			L3->next=(struct ListNode *)malloc(sizeof(struct ListNode));
			L3=L3->next;	
		}

	}
	return origin_L3;	

}

File: leetcode/021_Merge_Two_Sorted_Lists/main.java
package test;

public class main {
	
	public static void main(String[] args) { 
		ListNode L1;
		ListNode L2;
		ListNode L3;
		int L1_arr[]={1,2,4};
		int L2_arr[]={1,3,4};
		L1=addNodes(L1_arr);
		L2=addNodes(L2_arr);
		
		visitNodes(L1);
		visitNodes(L2);
		
		L3=mergeTwoLists(L1,L2);
		visitNodes(L3);
	}
	private static void visitNodes(ListNode node)
	{
		ListNode currNode=node;
		System.out.print(currNode.val+"->");
		while(currNode.next!=null)
		{
			currNode=currNode.next;
			System.out.print(currNode.val+"->");
		}
		System.out.println("");
		
	}
	private static ListNode addNodes(int []arr){
		
		if(arr.length==0)
			return null;
		
		ListNode currNode=new ListNode(arr[0]);
		ListNode head=currNode;
		currNode.next=null;
		
		for(int i=1;i<arr.length;i++){
			currNode.next=new ListNode(arr[i]);
			currNode=currNode.next;
		}//end of for
		currNode.next=null;
		return head;
	}
	public static ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode currL1=l1;
        ListNode currL2=l2;
        ListNode L3=null;
        ListNode origin_L3=null;
		int value=0;
		
		if (l1==null && l2==null)
			return null;
		
		L3=new ListNode(0);
		origin_L3=L3;
		
		while(currL1!=null || currL2!=null){
			if(currL2==null){
				value=currL1.val;
				if(currL1.next!=null){
					currL1=currL1.next;
				}else{
					currL1=null;
				}
			}else if(currL1==null){
				value=currL2.val;
				if(currL2.next!=null){
					currL2=currL2.next;
				}else{
					currL2=null;
				}
			}else if(currL1.val<=currL2.val){
				value=currL1.val;
				if(currL1.next!=null){
					currL1=currL1.next;
				}else{
					currL1=null;
				}
			}else{
				value=currL2.val;
				if(currL2.next!=null){
					currL2=currL2.next;
				}else{
					currL2=null;
				}
			}
			L3.val=value;
			
			if(currL1!=null || currL2!=null){
				L3.next=new ListNode(0);
				L3=L3.next;
			}
			
		}//end of while
		return origin_L3;
    }

	
}
 

    

